libs/Matchmaking/src/Peer/PeerDiscoveryAgent.cs

```diff
@@ -641,12 +645,20 @@ internal IDisposedEventDiscoveryTask StartDiscovery(string category)
 
         internal void Stop()
         {
+            // Do not wait for the end of the update thread since it runs external handlers,
+            // and we might be stuck for a long time/deadlock.
+            updateCts_.Cancel();
+            updateCts_.Dispose();
+
+            proto_.Stop();
```

Comment:
Is this the correct order?

Reply:
It shouldn't matter too much, should it?

Reply:
It's not clear if some tokens can still be in use until everything is stopped and all running threads are joined.

Reply:
It turns out that `ThrowIfCancellationRequested` will throw an ObjectDisposedException if the CTS is disposed (`IsCancellationRequested`, curiously, will not). This is kind of irrelevant in this case since in case of cancellation we want to throw anyway, but I made it cleaner anyway.